### PR TITLE
On method chain expression failure, look for missing method in earlier segments of the chain

### PR DESCRIPTION
**Pull request by iSwapna**
This PR tries to fix the issue: https://github.com/rust-lang/rust/issues/115222

As suggested by @estebank , I did the following:
1. Add new test `tests/ui/structs/method-chain-expression-failure.rs`
2. In `compiler/rusct_hir_tycheck/src/method/suggest.rs` 
   walking up the method chain and calling `probe_for_name` with the method name. But the call fails to return `Ok`.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -465,6 +465,42 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    .borrow()
+                    .expr_ty_adjusted_opt(rcvr_expr)
+                    .unwrap_or(Ty::new_misc_error(self.tcx)),
+            );
+            /* err.span_note(
+                span,
+                format!("TEST: the method `{item_name}` is implemented on `{rcvr_expr}`"),
+            ); */
+
+            if let Ok(_pick) = self.probe_for_name(
```
Comment (iSwapna) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:484]: Why is this `probe_for_name` not picking up B?

**Comment by oli-obk** [comment]
it looks like there are tests other than yours that are affected by this change. If you bless all ui tests, you should see some changes in your git status view

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:484]: what's the issue you're seeing exactly? What do you mean by `not picking up B`?

Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:484]: I am expecting the `probe_for_name` to find `struct B` as the receiver for the function `foo`. I am using the while loop to walk up the inheritance path to find the method `foo` in any of the predecessors. Is that making things clear?

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 2]
```diff
@@ -465,6 +465,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                rcvr_expr.hir_id,
+                ProbeScope::AllTraits,) {
+                    err.span_label(
+                        span,
+                        format!(
+                        "{item_kind} `{item_name}` found for this {:?}",
+                        prev_ty
+                        ),
+                 );
+                println!("Item name: {item_name} item_kind: {item_kind}");
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:500]:
```suggestion
```

File: tests/ui/typeck/issue-31173.stderr [hunk 3]
```diff
@@ -31,6 +31,7 @@ LL | |             found_e = true;
 LL | |             false
 LL | |         })
 LL | |         .cloned()
+   | |          -------- method `collect` found for this std::iter::TakeWhile<&mut std::vec::IntoIter<u8>, [closure@$DIR/issue-31173.rs:7:21: 7:25]>
```
Comment (estebank) [inline tests/ui/typeck/issue-31173.stderr:34]: Maybe we should to not show this label for this case, as it is somewhat redundant: the following label already imply that `collect` is available for `cloned`, except for the trait bounds. But then again it probably is ok either way.

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:485]: For debugging, you can do `err.span_node(span, format!("{prev_ty:?}"));` to unconditionally show what type each method call segment is found to be, in order to confirm that the logic before `probe_for_name` is correct. This all looks correct to me at first glance, but I wonder if there might be some subtle bug that we are missing.

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:485]: The thing I can think of is that maybe `expr_ty_adjusted_opt` is not finding a type for the `rcvr_expr`, but that's speculation.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 4]
```diff
@@ -476,28 +476,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
                         ),
                  );
-                break;
+                },
+                Err(MethodError::NoMatch(..)) => 
+                    println!("No Match"),
+                Err(MethodError::Ambiguity(..)) =>  println!("Ambiguity"),
+                Err(MethodError::PrivateMatch(..)) => println!("Private"),
+                Err(MethodError::IllegalSizedBound { .. }) => println!("IllegalSizedBound"),
+                Err(MethodError::BadReturnType) => println!("BadReturnType"),
```
Comment (iSwapna) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:505]: This outputs:
```
prev_ty: &C
item_name: foo
No Match
prev_ty: &B
item_name: foo
BadReturnType
prev_ty: &A
item_name: foo
No Match
```
So, what makes the return type of `B` bad?

File: tests/ui/typeck/issue-31173.stderr (hunk 3 above)
Comment (iSwapna) [reply tests/ui/typeck/issue-31173.stderr:34]: Let me know your decision. I do not understand the context of the typecheck yet!

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 4 above)
Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: @rustbot review

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: I'm pretty sure it's because of this check:

https://github.com/rust-lang/rust/blob/dac91a82e146d89fefe94dcbb53cd6c2662e354f/compiler/rustc_hir_typeck/src/method/probe.rs#L1697-L1704

Can you take a look at what those two types in the `debug!` are, as well as the actual value of the error? It might be that we might want to consider `BadReturnType` as "correct" for our purposes (if we don't change that earlier code that produces it, potentially adding a new more specific  `ProbeResult` variant).

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: Can we look at what the output of the following is?

https://github.com/rust-lang/rust/blob/dac91a82e146d89fefe94dcbb53cd6c2662e354f/compiler/rustc_hir_typeck/src/method/probe.rs#L1309

That might give us a clue at why we end up with a BadReturnType or NoMatch at the end of that function.

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: BTW, you can prepend your rustc invocation with `RUSTC_LOG` like `RUSTC_LOG=rustc_hir_typeck::method::probe=debug rustc foo.rs` to make these debug statements be printed out when executed.

Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: I was going to ask.. 
A few other questions:
How do I run `stage1` compiler on gdb? I tried but just pointing to the stage1 binary crashed in gdb. Do I need to build `stage2` as well and point to that? 

What do I do to run an `ignored` test? - I am currently editing to ensure a recompile

Is there a specific place in the documentation and code that could help me understand the guts of this piece of analysis?

**Comment by iSwapna** [comment]
Contents of tests/ui/structs/method-chain-expression-failure.stdout:
```
Applicable candidates: [(Candidate { xform_self_ty: &A, xform_ret_ty: Some(B), item: AssocItem { def_id: DefId(0:12 ~ method_chain_expression_failure[2e39]::{impl#0}::b), name:\
 "b", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
Match)]
Applicable candidates: [(Candidate { xform_self_ty: &B, xform_ret_ty: Some(C), item: AssocItem { def_id: DefId(0:14 ~ method_chain_expression_failure[2e39]::{impl#1}::c), name:\
 "c", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
Match)]
Applicable candidates: [(Candidate { xform_self_ty: &C, xform_ret_ty: Some(D), item: AssocItem { def_id: DefId(0:17 ~ method_chain_expression_failure[2e39]::{impl#2}::d), name:\
 "d", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
Match)]
prev_ty: &C
item_name: foo
No Match
prev_ty: &B
item_name: foo
self_ty=&B probe=Candidate { xform_self_ty: &B, xform_ret_ty: Some(()), item: AssocItem { def_id: DefId(0:15 ~ method_chain_expression_failure[2e39]::{impl#1}::foo), name: "foo\
", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }
1705 :: Consider_probe: comparing return_ty &B with xform ret ty ()
Error Type is Err(Sorts(ExpectedFound { expected: &B, found: () }))
Applicable candidates: [(Candidate { xform_self_ty: &B, xform_ret_ty: Some(()), item: AssocItem { def_id: DefId(0:15 ~ method_chain_expression_failure[2e39]::{impl#1}::foo), na\
me: "foo", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [\
] }, BadReturnType)]
BadReturnType
prev_ty: &A
item_name: foo
No Match

```

**Comment by iSwapna** [comment]
> @estebank Contents of tests/ui/structs/method-chain-expression-failure.stdout:
> 
> ```
> Applicable candidates: [(Candidate { xform_self_ty: &A, xform_ret_ty: Some(B), item: AssocItem { def_id: DefId(0:12 ~ method_chain_expression_failure[2e39]::{impl#0}::b), name:\
>  "b", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
> Match)]
> Applicable candidates: [(Candidate { xform_self_ty: &B, xform_ret_ty: Some(C), item: AssocItem { def_id: DefId(0:14 ~ method_chain_expression_failure[2e39]::{impl#1}::c), name:\
>  "c", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
> Match)]
> Applicable candidates: [(Candidate { xform_self_ty: &C, xform_ret_ty: Some(D), item: AssocItem { def_id: DefId(0:17 ~ method_chain_expression_failure[2e39]::{impl#2}::d), name:\
>  "d", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }, \
> Match)]
> prev_ty: &C
> item_name: foo
> No Match
> prev_ty: &B
> item_name: foo
> self_ty=&B probe=Candidate { xform_self_ty: &B, xform_ret_ty: Some(()), item: AssocItem { def_id: DefId(0:15 ~ method_chain_expression_failure[2e39]::{impl#1}::foo), name: "foo\
> ", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [] }
> 1705 :: Consider_probe: comparing return_ty &B with xform ret ty ()
> Error Type is Err(Sorts(ExpectedFound { expected: &B, found: () }))
> Applicable candidates: [(Candidate { xform_self_ty: &B, xform_ret_ty: Some(()), item: AssocItem { def_id: DefId(0:15 ~ method_chain_expression_failure[2e39]::{impl#1}::foo), na\
> me: "foo", kind: Fn, container: ImplContainer, trait_item_def_id: None, fn_has_self_parameter: true, opt_rpitit_info: None }, kind: InherentImplCandidate([], []), import_ids: [\
> ] }, BadReturnType)]
> BadReturnType
> prev_ty: &A
> item_name: foo
> No Match
> ```

In this case, should we check for return type of `foo()` to match the one implemented by `B` ? If so, we should first check if there is a variable the return value is getting assigned to? But I think, since this is a hint to the user, just matching by name and returning the first (closest to `D`) one as a possible match would be enough to point the user in the right direction?

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 4 above)
Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: @estebank Also wondering about the timeline expectations for PRs to be resolved - Do not want to take up all the time in learning on this PR if it's taking way too much of your time. But absolutely appreciate your time and look to learning more!

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:505]: > Is there a specific place in the documentation and code that could help me understand the guts of this piece of analysis?

In general, the best place to look at is the [dev guide](https://rustc-dev-guide.rust-lang.org/) for high level information, or the doc comments scattered throughout the different modules.

> What do I do to run an ignored test? - I am currently editing to ensure a recompile

What I do is [set up a rustup toolchain for my locally built](https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html?highlight=rustup#creating-a-rustup-toolchain) with the name `dev`, and after that calling `rustc +dev foo.rs` will compile `foo.rs`. Once you have that set up you can compile any of the tests (but this is only really useful for the tests that are self-contained, the ones that have dependencies are more annoying to deal with and it's easier to just run the tests for those).

What was the crash you saw with gdb? In general I rely heavily in debug statements instead of the debugger, but I've had good results with the lldb + vscode integration, so I don't know how to help you there.

> Also wondering about the timeline expectations for PRs to be resolved

There's no rush.

> Do not want to take up all the time in learning on this PR if it's taking way too much of your time. But absolutely appreciate your time and look to learning more!

Don't worry, and it's my pleasure. We all started somewhere.

**Comment by estebank** [comment]
> But I think, since this is a hint to the user, just matching by name and returning the first (closest to D) one as a possible match would be enough to point the user in the right direction?

If we can filter the method with the right return type, I would prefer to suggest it, otherwise I would mention all of them. Give me a minute to check the rest of the PR and correlate with the logs you have here.

**Comment by estebank** [comment]
Ah! I see what's happening. `A.b().c().d().foo();` as written of course fails, but the solution *in this case* is to write `A.b().foo();`, but the logic *as written* is (somewhat incorrectly) trying to figure out if `A.b().foo().c().d();` is valid. The former is a case of additional elements in the chain, while the later is a case of misplaced method call. I think we can support *both* with high accuracy. For the latter the current logic we need to make a change: if you look at `return_ty` being `&B` it means that it's looking for a method that will return `prev_ty`, not just one that has that as `Self`. I'll point it out inline.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 5]
```diff
@@ -465,6 +465,48 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    .unwrap_or(Ty::new_misc_error(self.tcx)),
+            );
+            err.span_note(span, format!("{prev_ty:?}"));
+            println!("prev_ty: {}", prev_ty );
+            println!("item_name: {}", item_name);
+
+            match self.probe_for_name(
+            Mode::MethodCall,
+                item_name,
+                Some(prev_ty),
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:486]: This is the problem, this should either be a type that has a valid `.c()` method *or* `None`, but by passing it here we're filtering on all methods that conform to `<T as _>::item_name(&T) -> &T`, which is why we were getting `BadReturnType`. Also, there's `probe_for_name_many` that can give you back multiple methods that might apply and it's more likely what we would want to be using here.

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:486]: Thinking this further, you *want* to pass `None` in 486, but use `probe_for_name_many` instead. This way once we get the result we can look at all the potential places the options could go to suggest an accurate code change, and if none are *positively* valid we emit a `span_note` or `span_label` pointing at the place where the call could go.

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:486]: To be clear, I'm ok with having this PR cover only the "fallback" case and only start providing structured suggestions in a subsequent PR.

Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:486]: Been busy at work and will get to this over weekend. Exciting to get to the guts. Thanks for all the pointers!

Comment (iSwapna) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:486]: @estebank Spent a few hours on this - I do not understand how `return_type::None` would work to return any function with `foo` `item_name`. Or are we only looking for `foo()` which has a return type of `None`. Are we trying to set `return_type` (for `probe_for_name_namy`) to be the `return_type` of the method with `item_name` in the user's incorrect code : `A.b().c().d().foo()` ?

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:486]: Calling `self.probe_for_name_many(Mode::MethodCall, item_name, None,IsSuggestion(true), prev_ty, rcvr_expr.hir_id, ProbeScope::AllTraits)` instead of `self.probe_for_name_many(Mode::MethodCall, item_name, Some(prev_ty), IsSuggestion(true), prev_ty, rcvr_expr.hir_id, ProbeScope::AllTraits)` will make it so that the machinery looks for methods on all traits of name `item_name` (in this case `foo`) that can be named on `prev_ty`, regardless of return type (that's the `None` we're passing in now).

Before we were explicitly asking for the return type to also be `prev_ty`, which filters on all methods on `prev_ty` that also return `prev_ty`, which there were none and why we were ending up with `BarReturnType`.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 6]
```diff
@@ -465,6 +465,43 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        for matched_method in &all_methods {
+            //err.span_note(sp: matched_method, msg:format!("Possible match: {matched_method:?}"));
+            println!("At the span of the matched_method: {matched_method:?}");
+
+        }
+
+
+
+    }
+
```
Comment (iSwapna) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:504]:
```suggestion
            println!("At the span of the matched_method: {matched_method:?}");

        }
    }

```

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 7]
```diff
@@ -465,6 +465,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        // Collect all possible candidates in the chain with same 'item_name'
+        let mut all_methods: Vec<super::probe::Candidate<'_>> = vec![];
+        if let Mode::MethodCall = mode && let SelfSource::MethodCall(mut rcvr_expr) = source {
+        while let hir::ExprKind::MethodCall(_path_segment, parent_expr, _args, _span) =
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:471]: You will likely want to keep around wither the span or the path_segment.span for the final error, so you can point at the right place in the user's expression when talking about the method being available, because there might be more than one chain in the method call that have the method available.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 8]
```diff
@@ -465,6 +465,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                rcvr_expr.hir_id,
+                ProbeScope::AllTraits,));
+        }
+        // Check if any of the candidates match
+        // 1. If one or more match both the parameters and return type, emit a span_note at the level of the candidate of a possible match
+        // 2. If none match per case 1. but there are name matches, (ie., the 'item_name' matches one or more candidates), emit a span note with a suggestion
+
+        for matched_method in &all_methods {
+            //err.span_note(sp: matched_method, msg:format!("Possible match: {matched_method:?}"));
+            println!("At the span of the matched_method: {matched_method:?}");
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:497]: `tcx.def_span(matched_method.item.def_id)` will give you the span of where the method is defined and `tcx.def_path_str(matched_method.item.def_id)` will give you a `String` with the `path::to::Type::method` that you would want to present to the user in the message.

We already know that the `Self` type is correct (`probe_for_name` will take care of that). The only thing left might be to try and do some check with `self.can_eq(type_after_call, matched_method.xform_ret_ty.unwrap())`, but we can likely leave that for a later PR.

File: tests/ui/structs/method-chain-expression-failure.stderr [hunk 9]
```diff
@@ -0,0 +1,12 @@
+error[E0599]: no method named `foo` found for struct `D` in the current scope
+  --> $DIR/method-chain-expression-failure.rs:24:19
+   |
+LL | struct D;
+   | -------- method `foo` not found for this struct
+...
+LL |     A.b().c().d().foo();
+   |                   ^^^ method not found in `D`
```
Comment (estebank) [inline tests/ui/structs/method-chain-expression-failure.stderr:8]: Assuming that both B and C have a foo method, I think the ideal output to strive for would be something like

```
error[E0599]: no method named `foo` found for struct `D` in the current scope
  --> $DIR/method-chain-expression-failure.rs:24:19
   |
LL | struct D;
   | -------- method `foo` not found for this struct
...
LL |     A.b().c().d().foo();
   |           --- --- ^^^ method not found in `D`
   |           |   |
   |           |   method `foo` is available on `C`
   |           method `foo` is available on `B`
note: method `foo` is available on `B` and `C`
   |
   |     fn foo(&self) {}
   |        ^^^ method available on `B` here
...
   |     fn foo(&self) {}
   |        ^^^ method available on `C` here
help: you might have meant to call `foo` earlier in the method chain
   |
LL -     A.b().c().d().foo();
LL +     A.b().c().foo();
   |
LL -     A.b().c().d().foo();
LL +     A.b().foo();
   |
```

Going out of our way we would also check if `.d()` could be called after `.foo()` and be more accurate with that, but again, that's not something we need to accomplish all at once. I'm more than happy if all we have is the new span_labels on the main message.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 10]
```diff
@@ -465,6 +465,43 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        // Collect all possible candidates in the chain with same 'item_name'
+        let mut all_methods: Vec<super::probe::Candidate<'_>> = vec![];
+        if let Mode::MethodCall = mode && let SelfSource::MethodCall(mut rcvr_expr) = source {
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:470]: A trick for later: rustfmt today ignores the entire expression when it doesn't know what to do, and it doesn't know what to do with `if let` chains. You can temporarily change this to `if true {`, run rustfmt so that the body is properly formatted, and revert this line to what it is today. This will make it easier to ensure the code is properly formatted when running `./x.py fmt` without as much effort.

File: tests/ui/issues/issue-39175.stderr [hunk 11]
```diff
@@ -4,6 +4,8 @@ error[E0599]: no method named `exec` found for mutable reference `&mut Command`
 LL |     Command::new("echo").arg("hello").exec();
    |                                       ^^^^ method not found in `&mut Command`
    |
+note: Possible match: "exec"
+  --> $SRC_DIR/std/src/os/unix/process.rs:LL:COL
    = help: items from traits can only be used if the trait is in scope
 help: the following trait is implemented but not in scope; perhaps add a `use` for it:
```
[the remark this answers is not in the document]
Comment (iSwapna) [reply tests/ui/issues/issue-39175.stderr:10]: I think I understand - will try skipping the first

File: tests/ui/structs/method-chain-expression-failure.stderr [hunk 12]
```diff
@@ -0,0 +1,17 @@
+error[E0599]: no method named `foo` found for struct `D` in the current scope
+  --> $DIR/method-chain-expression-failure.rs:24:19
+   |
+LL | struct D;
+   | -------- method `foo` not found for this struct
+...
+LL |     fn foo(&self) {}
+   |     ------------- method available on `&C` here
```
Comment (iSwapna) [inline tests/ui/structs/method-chain-expression-failure.stderr:8]: @estebank  How do I remove the `&` and the method level comments appear to come ahead (I think since that's the src code order). Is that ok?

File: tests/ui/structs/method-chain-expression-failure.stderr (hunk 9 above)
Comment (iSwapna) [reply tests/ui/structs/method-chain-expression-failure.stderr]: @estebank How do I add the `LL-`,  `LL+` comments? I think I can collect all the possible candidates and create the `note` and `help`

File: tests/ui/structs/method-chain-expression-failure.stderr (hunk 12 above)
Comment (estebank) [reply tests/ui/structs/method-chain-expression-failure.stderr:8]: Ordering is ok, we can side-step that issue by introducing a `note` instead.

There's a `peel_refs` argument on `ty::Ty` IIRC, which would let you do that, but it is fine to leave it as is as well.

File: tests/ui/issues/issue-39175.stderr (hunk 11 above)
[the remark this answers is not in the document]
Comment (estebank) [reply tests/ui/issues/issue-39175.stderr:10]: You can accomplish that with something like

```rust
let mut iter = path.segments.iter().rev();
iter.next();
while let Some(segment) = iter.next() {
    ...
}
```

[the remark this answers is not in the document]
Comment (iSwapna) [reply tests/ui/issues/issue-39175.stderr:10]: How do I get the path?

**Comment by iSwapna** [comment]
@estebank Few issues:
1. Not sure how to get the path from `Expr` 
2. Limiting the `ProbeScope` to  `TraitsInScope` did not help- not sure why?
3. Rebasing on master raised a lot of conflicts with file paths being changed. What is the protocol to merge them? Do I hand merge or close this pull request and open a new one from the head of master?

**Comment by estebank** [comment]
> Not sure how to get the path from Expr

Ignore the recommendation, you have a `while let` with pattern matching, not a `vec![PathSegment]`, so you'll just need an `if` check whether you're in the first iteration, that's all.

> Limiting the ProbeScope to TraitsInScope did not help- not sure why?

I'll have to look at it again.

> Rebasing on master raised a lot of conflicts with file paths being changed. What is the protocol to merge them? Do I hand merge or close this pull request and open a new one from the head of master?

What I usually end up doing is `git rebase master -i` and squash every commit to make the merge easier to deal with on one go. We continue on the same PR, in general.

File: tests/ui/suggestions/chain-method-call-mutation-in-place.stderr [hunk 13]
```diff
@@ -18,7 +18,10 @@ error[E0599]: no method named `sort` found for unit type `()` in the current sco
   --> $DIR/chain-method-call-mutation-in-place.rs:3:72
    |
 LL |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
-   |                                                                        ^^^^ method not found in `()`
+   |                   -----------                       ------------------ ^^^^ method not found in `()`
+   |                   |                                 |
+   |                   |                                 method `sort` is available on `&mut [i32]`
+   |                   method `sort` is available on `Vec<i32>`
```
Comment (iSwapna) [inline tests/ui/suggestions/chain-method-call-mutation-in-place.stderr:24]: @estebank how do I get rid of this pesky one? Sort is available on `Vec` but since `sort_by_key`s return is unit, the check up method chain seems irrelevant.

Comment (estebank) [reply tests/ui/suggestions/chain-method-call-mutation-in-place.stderr:24]: The spans here seem to be shifted by one. The first span should be pointing at the `vec![1, 2, 3]` and the second one to the `.collect::<Vec<i32>>()`. The logic to identify the call is correct, you just likely need to point at the receiver of the call, not the segment.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 14]
```diff
@@ -1240,7 +1239,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+
+                for _matched_method in self.probe_for_name_many(
+                    Mode::MethodCall,
+                    item_name,
+                    None,
+                    IsSuggestion(true),
+                    prev_ty,
+                    rcvr_expr.hir_id,
+                    ProbeScope::TraitsInScope,) {
+                        err.span_label(method_span, format!("{item_kind} `{item_name}` is available on `{prev_ty}`"));
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1262]: If you use `parent_expr.span`, it would make the error be correct, but when you have more than one element in the chain you would end up with overlapping spans which won't look great.

An option we have is to iterate once using the `while let` as you have it now, but collect the messages and spans in a vec and emit the diagnostics *afterwards* by iterating over the vec and *peeking* at the next span.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 15]
```diff
@@ -1240,7 +1239,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
         }
+        // Only if an appropriate error source is not found, check method chain for possible candiates
+        let error_source_not_found = unsatisfied_predicates.is_empty();
+        if error_source_not_found && let Mode::MethodCall = mode && let SelfSource::MethodCall(mut rcvr_expr) = source {
+        while let hir::ExprKind::MethodCall(_path_segment, parent_expr, _args, method_span) =
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1245]: I would prefer we called `parent_expr` to something that includes `rcvr`/`receiver` in the name, because the `parent` concept is highly overloaded :)

File: tests/ui/structs/method-chain-expression-failure.stderr [hunk 16]
```diff
@@ -0,0 +1,14 @@
+error[E0599]: no method named `foo` found for struct `D` in the current scope
+  --> $DIR/method-chain-expression-failure.rs:24:19
+   |
+LL | struct D;
+   | -------- method `foo` not found for this struct
+...
+LL |     A.b().c().d().foo();
+   |           ---     ^^^ method not found in `D`
+   |           |
+   |           method `foo` is available on `&B`
```
Comment (estebank) [inline tests/ui/structs/method-chain-expression-failure.stderr:10]: To drive the point home, if you look at the impl the span label should be pointing at the `.b()` which returns `B` and not at the method call `.c()` which returns `C`.

**Review by estebank — COMMENTED** [review]
I love the results so far! On your next change, please squash your commits as I don't think we need the prior attempts anymore and expect us to land this shortly.

**Comment by estebank** [comment]
You'll need to run `x.py test tests/ui --bless` again. Let's also squash the remaining commits and I think we can merge the PR after that.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 17]
```diff
@@ -1240,7 +1239,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        .unwrap_or(Ty::new_misc_error(self.tcx)),);
+
+                for _matched_method in self.probe_for_name_many(
+                    Mode::MethodCall,
+                    item_name,
+                    None,
+                    IsSuggestion(true),
+                    rcvr_ty,
+                    source_expr.hir_id,
+                    ProbeScope::TraitsInScope,) {
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1261]:
```suggestion
                    ProbeScope::TraitsInScope,
                ) {
```

**Comment by iSwapna** [comment]
> You'll need to run `x.py test tests/ui --bless` again. Let's also squash the remaining commits and I think we can merge the PR after that.

Had found an issue. Done now.
